aptos-move/framework/aptos-framework/sources/storage_gas.move

```diff
@@ -0,0 +1,359 @@
+    }
+
+    fun validate_curve(curve: &GasCurve) {
+        assert!(curve.max_gas >= curve.min_gas, error::invalid_argument(EINVALID_GAS_RANGE));
+        let points = &curve.points;
+        let len = vector::length(points);
+        let i = 0;
+        while (i <= len) {
+            let cur = if (i == 0) { &Point {x: 0, y: 0} } else { vector::borrow(points, i - 1) };
+            let next = if (i == len) { &Point {x: BASIS_POINT_DENOMINATION, y: BASIS_POINT_DENOMINATION} } else { vector::borrow(points, i) };
+            assert!(cur.x < next.x && cur.y <= next.y, error::invalid_argument(EINVALID_MONOTONICALLY_NON_DECREASING_CURVE));
+            i = i + 1;
+        }
+    }
+
+    fun calculate_gas(max_usage: u64, current_usage: u64, curve: &GasCurve): u64 {
+        let capped_current_usage = if (current_usage > max_usage) max_usage else current_usage;
+        let points = &curve.points;
+        let num_points = vector::length(points);
+        let current_usage_bps = capped_current_usage * BASIS_POINT_DENOMINATION / max_usage;
+
+        // Check the corner case that current_usage_bps drops before the first point.
+        let (left, right) = if (num_points == 0) {
+            (&Point {x: 0, y: 0}, &Point {x: BASIS_POINT_DENOMINATION, y: BASIS_POINT_DENOMINATION})
+        } else {
+            let (i, j) = (0, num_points - 1);
+            while (i < j) {
+                let mid = j - (j - i) / 2;
+                if (current_usage_bps < vector::borrow(points, mid).x) {
+                    j = mid - 1;
+                } else {
+                    i = mid;
+                };
+            };
+            if (current_usage_bps < vector::borrow(points, 0).x) {
+                (&Point {x: 0, y: 0}, vector::borrow(points, 0))
+            } else {
+                // Check the corner case that current_usage_bps drops after the last point.
+                (
+                    vector::borrow(points, i),
+                    if (i == num_points - 1) {
+                        &Point { x: BASIS_POINT_DENOMINATION, y: BASIS_POINT_DENOMINATION }
+                    } else {
+                        vector::borrow(points, i + 1)
+                    }
+                )
+            }
+        };
+
+        curve.min_gas + (curve.max_gas - curve.min_gas) * (left.y + (current_usage_bps - left.x) * (right.y - left.y) / (right.x - left.x)) / BASIS_POINT_DENOMINATION
+    }
+
+    fun calculate_read_gas(config: &UsageGasConfig, usage: u64): u64 {
+        calculate_gas(config.target_usage, usage, &config.read_curve)
+    }
+
+    fun calculate_create_gas(config: &UsageGasConfig, usage: u64): u64 {
+        calculate_gas(config.target_usage, usage, &config.create_curve)
+    }
+
+    fun calculate_write_gas(config: &UsageGasConfig, usage: u64): u64 {
+        calculate_gas(config.target_usage, usage, &config.write_curve)
+    }
+
+    public(friend) fun on_reconfig() acquires StorageGas, StorageGasConfig {
```

Comment:
Can this fail? If this fails, it crashes the network

Reply:
all the code triggered by reconfig will crash the network if fails. I think we avoid all the possible overflow issues. But if you can anything risky, I am glad to discuss.